### PR TITLE
tools/makeqstrdefs.py: Run qstr preprocessing in parallel.

### PR DESCRIPTION
**Pull request by jimmo**
This gives a substantial speedup of the preprocessing step (i.e. qstr.i.last).
For example on a clean build, making qstr.i.last:

21s -> 4s on STM32 (WB55)
8.9 -> 1.8s on Unix (dev).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

### CODE REVIEW

**Comment by stinos** [comment]
Thus is probably worth it, even though the code becomes way harder to read. Why is the tempfile approach needed?

File: py/makeqstrdefs.py [hunk 1]
```diff
@@ -39,11 +44,28 @@ def preprocess():
...
+        for flags, sources in ((args.cflags, csources), (args.cxxflags, cxxsources),):
+            if not sources:
+                continue
+            batch_size = max(10, len(sources) // _PREPROCESS_PARALLELISM)
+            for i in range(_PREPROCESS_PARALLELISM):
+                first, last = i * batch_size, min(len(sources), (i+1)*batch_size)
+                (fd, _path,) = tempfile.mkstemp()
+                process = subprocess.Popen(args=args.pp + flags + sources[first:last], stdout=fd)
+                tasks.append((fd, process))
+                if last >= len(sources):
```
Comment (stinos) [inline py/makeqstrdefs.py:59]: `==` instead of `>=` ?

**Comment by dpgeorge** [comment]
Fantastic!  This is not something that could have been expressed in a Makefile...

But it seems to be failing on a lot of ports, maybe it misses processing some files?

**Comment by jimmo** [comment]
> But it seems to be failing on a lot of ports, maybe it misses processing some files?

True story. Just pushed a fixed.

**Comment by jimmo** [comment]
Final issue was python2 compatibility. Fixed.

**Comment by jimmo** [comment]
Updated to use an alternative implementation by @stinos using `multiprocessing.dummy.Pool` which is much simpler and easier to read.

File: py/makeqstrdefs.py [hunk 2]
```diff
@@ -20,6 +21,11 @@
 # Extract MP_COMPRESSED_ROM_TEXT("") macros.  (Which come from MP_ERROR_TEXT)
 _MODE_COMPRESS = "compress"
 
+# How many concurrent pre-processor instances to run to extract QSTRs.
+_PREPROCESS_PARALLELISM = 8
```
Comment (dlech) [inline py/makeqstrdefs.py:25]: Should we make this depend on number of CPUs? Something like...

    _PREPROCESS_PARALLELISM  = multiprocessing.cpu_count() - 1

File: py/makeqstrdefs.py [hunk 3]
```diff
@@ -20,6 +21,11 @@
...
+# How many files to pre-process at a time.
+_PREPROCESS_BATCH_SIZE = 20
```
Comment (dpgeorge) [inline py/makeqstrdefs.py:27]: Why not just do the maximum, ie each process has `TOTAL_FILES / NUM_PROCESS`?

File: py/makeqstrdefs.py (hunk 2 above)
Comment (jimmo) [reply py/makeqstrdefs.py]: Done

File: py/makeqstrdefs.py (hunk 3 above)
Comment (jimmo) [reply py/makeqstrdefs.py]: Done

File: py/makeqstrdefs.py [hunk 4]
```diff
@@ -39,11 +40,26 @@ def preprocess():
...
+    cpus = multiprocessing.cpu_count()
+    p = multiprocessing.dummy.Pool(cpus)
+    with open(args.output[0], "wb") as out_file:
+        for flags, sources in (
+            (args.cflags, csources),
+            (args.cxxflags, cxxsources),
+        ):
+            if not sources:
+                continue
+            batch_size = (len(sources) + cpus - 1) // cpus
```
Comment (thyrrestrup) [inline py/makeqstrdefs.py:59]: Is this batch_size correct?

File: py/makeqstrdefs.py [hunk 5]
```diff
@@ -39,11 +40,26 @@ def preprocess():
...
-        if cxxsources:
-            subprocess.check_call(args.pp + args.cxxflags + cxxsources, stdout=out_file)
+
+    def pp(flags):
+        def run(files):
+            return subprocess.check_output(args.pp + flags + files)
+
+        return run
+
+    cpus = multiprocessing.cpu_count()
```
Comment (thyrrestrup) [inline py/makeqstrdefs.py:50]: Should this check for the not implemented error? and make sure we have at least 1 cpu.

```
cpus = 1
try:
    cpus = multiprocessing.cpu_count()
except NotImplementedError:
    pass

if cpus > 1:
    cpus -= 1
```

File: py/makeqstrdefs.py [hunk 6]
```diff
@@ -39,11 +40,26 @@ def preprocess():
...
+    p = multiprocessing.dummy.Pool(cpus)
+    with open(args.output[0], "wb") as out_file:
+        for flags, sources in (
+            (args.cflags, csources),
+            (args.cxxflags, cxxsources),
+        ):
+            if not sources:
```
Comment (stinos) [inline py/makeqstrdefs.py:57]: Shouldn't be needed? chunks will be empty as well if source is empty and I assume `imap` will also not do anything if there's no input

File: py/makeqstrdefs.py (hunk 4 above)
Comment (stinos) [inline py/makeqstrdefs.py:59]: isn't this like divmod or something else from math module?

Comment (jimmo) [reply py/makeqstrdefs.py:59]: @stinos it's definitely not builtin `divmod` which is `(a // b, a % b)`. It's the "ceiling integer division". To me this is the idiomatic way to express this and I'm not aware of any other Python standard library or builtin that provides this (although that would be nice if there was such a thing). Open to suggestions!

File: py/makeqstrdefs.py (hunk 6 above)
Comment (jimmo) [reply py/makeqstrdefs.py:57]: @stinos when `sources` is empty, `batch_size` is zero, and you can't do `range(x, y, 0)`.

I could solve this by making a minimum batch size of 1 or something but the `if not sources:` seems clearer to me.

File: py/makeqstrdefs.py (hunk 4 above)
Comment (peterhinch) [reply py/makeqstrdefs.py:59]: > ceiling integer division

I think you'll find it's floor integer division

File: py/makeqstrdefs.py (hunk 5 above)
Comment (jimmo) [reply py/makeqstrdefs.py]: True (although this is supported just about everywhere, including Windows, Linux, Mac, *BSD). multiprocessing.cpu_count() is a wrapper around os.cpu_count() anyway, so implemented as

`cpus = os.cpu_count() or 1`

File: py/makeqstrdefs.py (hunk 6 above)
Comment (stinos) [reply py/makeqstrdefs.py:57]: For me it's the opposite actually :) looking at the line calculating batch_size I really can't be bothered figuring out what it does for each possible number, so I see a calculation which for all i know could return 0, which makes me wondering what's going to happen in that case.

File: py/makeqstrdefs.py (hunk 4 above)
Comment (stinos) [reply py/makeqstrdefs.py:59]: `math.ceil(len(sources)/cpus) or 1` then ?

Comment (dpgeorge) [reply py/makeqstrdefs.py:60]: best not to introduce floats in to this calculation which is conceptually integer based

**Comment by dpgeorge** [comment]
Merged in a7932ae4e66bd1c354eee357f18393bd83144144

**Pull request closed without merging**